### ISSUE TITLE
[labs/observers] Consider adding polyfill for ResizeObserver

### ISSUE THREAD

**Issue by steverep** [lit/lit#3622]
### Should this be an RFC?

- [X] This is not a substantial change

### Which package is this a feature request for?

Observers (@lit-labs/observers)

### Description

Currently, the controller for the `ResizeObserver` simply logs a console warning when a browser appears not to support the API, and there's no way to alter this behavior.  This means that, in order to load a polyfill dynamically, you need to hook into `connectedCallback` anyway, which defeats a chunk of the convenience the `observers` package provides.

It would be nice if the package either:
1. Builds in polyfill support (similar to the virtualizer package), or
2. Provides an option to pass a function or other means to dynamically import a user-defined polyfill.

### Alternatives and Workarounds

It's possible to either:
1. Load a polyfill in `connectedCallback`, which has the disadvantages described in the previous section, or
2. Load a polyfill globally, which may override the polyfill inside other packages (e.g. virtualizer)

**Comment by justinfagnani** [comment]
I thought we were going to _remove_ the ResizeObserver from virtualizer. @usergenic @graynorton ?

Including polyfills in libraries is usually a bad idea. It ends up impacting apps that support only browsers that support the feature, and it bakes in a specific polyfill when an app may want to choose a different one, or other libraries may bundle one.

Loading the polyfill globally is generally the right approach. Only the app knows the browser environments it's deployging to.

**Comment by steverep** [comment]
The code in the virtualizer comments that the polyfill was "lightly modified", so perhaps it was only included because the released version didn't work for some reason?  The only one I see commonly used is https://github.com/que-etc/resize-observer-polyfill.

The global approach is fine I think.  Perhaps the only reason we do it dynamically is not to collide with virtualizer (cc @bramkragten).

**Comment by usergenic** [comment]
For the record, yes we do plan on *removing* the ResizeObserver polyfill from Virtualizer, and I think we'll try to get that done relatively soon.  The one we provide is currently a tweaked version that I believe enabled more reliable observation through layers of Shadow DOM etc.  I don't know if it is vended elsewhere in a ready-to-go form yet, but we'll ensure that there's an option before we remove it.

**Comment by graynorton** [comment]
Yes, the history here is that neither of the two most widely used polyfills properly supported detection within shadow roots, and the one that performed better in our tests and was most straightforward to patch was essentially unmaintained.

This is why we ended up forking the polyfill and shipping it with the library; we weren't interested in vending and supporting a general-purpose ResizeObserver polyfill, but we needed it for IE support. We came up with a mechanism for auto-loading it that added minimal weight to the library itself and didn't affect anything outside of Virtualizer, but with native support now at close to 100% due to the declining usage of IE11 and legacy Edge, we've decided to remove the auto-loading functionality and instead direct developers who need to support these older browsers to conditionally load it themselves.

@usergenic @justinfagnani I actually don't see much upside in removing the polyfill from the Virtualizer package altogether as long as IE11 and old Edge are still in Lit's official support matrix. A logical time to remove would be when the support matrix changes, presumably with a Lit 3.0.

**Comment by justinfagnani** [comment]
Thanks for the info @graynorton !

The more things that pop up that are constrained by IE support (WeakRef use in `lit-labs/context` for example), the more I want to just remove IE support altogether sooner rather than later.

**Comment by justinfagnani** [comment]
Since we've removed IE support and ResizeObserver has been in baseline since 2020, I'm going to close this an open an issue on removing the polyfill from virtualizer.
